packages/legend-application-query/src/components/Core_LegendQueryApplicationPlugin.tsx

```diff
@@ -725,4 +764,274 @@ export class Core_LegendQueryApplicationPlugin extends LegendQueryApplicationPlu
+      ): ((editorStore: QueryEditorStore) => GeneratorFn<void>) | undefined => {
+        function* buildGraph(): GeneratorFn<void> {
+          // do nothing
+        }
+        if (editorStore instanceof ExistingQueryEditorStore) {
+          const query = editorStore.query;
+          const dataSpaceTaggedValue = query?.taggedValues?.find(
+            (taggedValue) =>
+              taggedValue.profile === QUERY_PROFILE_PATH &&
+              taggedValue.tag === QUERY_PROFILE_TAG_DATA_SPACE &&
+              isValidFullPath(taggedValue.value),
+          );
+          if (
+            dataSpaceTaggedValue ||
+            query?.executionContext instanceof QueryDataSpaceExecutionContext
+          ) {
+            return buildGraph;
+          }
+        }
+        return undefined;
+      },
+    ];
+  }
+
+  getExtraQueryBuilderPropagateExecutionContextChangeHelper?(): QueryBuilderPropagateExecutionContextChangeHelper[] {
+    return [
+      (
+        queryBuilderState: QueryBuilderState,
+        isGraphBuildingNotRequired?: boolean,
+      ): (() => Promise<void>) | undefined => {
+        /**
+         * Propagation after changing the execution context:
+         * - The mapping will be updated to the mapping of the execution context
+         * - The runtime will be updated to the default runtime of the execution context
+         * - If no class is chosen, try to choose a compatible class
+         * - If the chosen class is compatible with the new selected execution context mapping, do nothing, otherwise, try to choose a compatible class
+         */
+        const propagateExecutionContextChange = async (): Promise<void> => {
+          const dataSpaceQueryBuilderState = guaranteeType(
+            queryBuilderState,
+            DataSpaceQueryBuilderState,
+          );
+          const mapping =
+            dataSpaceQueryBuilderState.executionContext.mapping.value;
+          const mappingModelCoverageAnalysisResult =
+            dataSpaceQueryBuilderState.dataSpaceAnalysisResult?.mappingToMappingCoverageResult?.get(
+              mapping.path,
+            );
+          const editorStore = (
+            queryBuilderState.workflowState
+              .actionConfig as QueryBuilderActionConfig_QueryApplication
+          ).editorStore;
+          if (
+            dataSpaceQueryBuilderState.dataSpaceAnalysisResult &&
+            mappingModelCoverageAnalysisResult
+          ) {
+            if (
+              !isGraphBuildingNotRequired &&
+              dataSpaceQueryBuilderState.isLightGraphEnabled
+            ) {
+              const supportBuildMinimalGraph =
+                editorStore.applicationStore.config.options
+                  .TEMPORARY__enableMinimalGraph;
+              if (
+                editorStore.enableMinialGraphForDataSpaceLoadingPerformance &&
+                supportBuildMinimalGraph
+              ) {
+                try {
+                  const stopWatch = new StopWatch();
```

Comment:
additionally below code is repeated. 
We should have a buildDataspaceMinialGraph in `V1_DSL_DataSpace_PureGraphManagerExtension` that takes in minimalEntities, dataspaceInfo, executionKeyToBuild, 
and builds the required graph. I think we already have this logic in `V1_DSL_DataSpace_PureGraphManagerExtension`

Reply:
Most of these lines are for building entities. In terms of building graph, there is only one line `await dataSpaceQueryBuilderState.graphManagerState.graphManager.buildGraph(..)`
  

For the function in `V1_DSL_DataSpace_PureGraphManagerExtension `, 
these two look similar but one takes in V1_ DataSpaceAnalysisResult,  another takes in DataSpaceAnalysisResult. 
I find it will create more chaos if I abstract logic of building entities and then building graph to a new standalone function.